### PR TITLE
Mon 4712 ad

### PR DESCRIPTION
**Pull request by bouda1**
# Pull Request Template

## Description

* Check results are smaller and faster now. We can even improve them, but this will come later.
* We have a new service that is anomalydetection.
* Several fixes also come with this PR.

**Fixes** # (issue)

## Type of change

- [X] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [X] 20.04.x (master)

### CODE REVIEW

File: modules/external_commands/src/processing.cc [hunk 1]
```diff
@@ -571,7 +578,7 @@ processing::processing()
 
 processing::~processing() throw() {}
 
-bool processing::execute(std::string const& cmdstr) const {
+bool processing::execute(const std::string& cmdstr) const {
```
Comment (SylvestreG) [inline modules/external_commands/src/processing.cc]: the const is not at the right place ;)

File: src/broker.cc [hunk 2]
```diff
@@ -1079,7 +1079,7 @@ int broker_host_check(int type,
                       int timeout,
                       int early_timeout,
                       int retcode,
-                      char* cmdline,
+                      const char* cmdline,
```
Comment (SylvestreG) [inline src/broker.cc]: char const* ?

File: src/main.cc [hunk 3]
```diff
@@ -93,6 +94,9 @@ int main(int argc, char* argv[]) {
   // Load singletons and global variable.
   config = new configuration::state;
 
+  // Hack to instanciate the logger.
+  configuration::applier::logging::instance();
+
```
Comment (SylvestreG) [inline src/main.cc]: do you want to keep this ?

File: src/simumod/coder.pl [hunk 4]
```diff
@@ -115,15 +115,15 @@
...
-}
+//static std::string timeval_str(timeval const& t) {
+//  struct tm* nowtm;
+//  time_t nowtime = t.tv_sec;
+//  char tmbuf[64], buf[64];
+//  nowtm = localtime(&nowtime);
+//  strftime(tmbuf, sizeof tmbuf, "%Y-%m-%d %H:%M:%S", nowtm);
+//  snprintf(buf, sizeof buf, "%s.%06ld", tmbuf, t.tv_usec);
+//  return std::string(buf);
+//}
```
Comment (SylvestreG) [inline src/simumod/coder.pl]: you can remove this code

File: src/xpddefault.cc [hunk 5]
```diff
@@ -181,6 +186,7 @@ int xpddefault_initialize_performance_data() {
 int xpddefault_cleanup_performance_data() {
   // free memory.
   delete[] xpddefault_host_perfdata_file_template;
+  std::cout << "cleanup xpddefault_service_perfdata_file_template\n";
```
Comment (SylvestreG) [inline src/xpddefault.cc]: leftovers

File: src/xpddefault.cc [hunk 6]
```diff
@@ -76,11 +77,15 @@ int xpddefault_initialize_performance_data() {
...
       string::dup(config->service_perfdata_file_template());
+  std::cout << "xpddefault_initialize_performance_data() :"
+    << config->service_perfdata_file_template() << "\n";
 
   // process special chars in templates.
   xpddefault_preprocess_file_templates(xpddefault_host_perfdata_file_template);
   xpddefault_preprocess_file_templates(
       xpddefault_service_perfdata_file_template);
+  std::cout << "process special chars in xpddefault_service_perfdata_file_template:"
+    << xpddefault_service_perfdata_file_template << "\n";
```
Comment (SylvestreG) [inline src/xpddefault.cc]: leftovers

File: src/xpddefault.cc [hunk 7]
```diff
@@ -76,11 +77,15 @@ int xpddefault_initialize_performance_data() {
       string::dup(config->host_perfdata_file_template());
   xpddefault_service_perfdata_file_template =
       string::dup(config->service_perfdata_file_template());
+  std::cout << "xpddefault_initialize_performance_data() :"
+    << config->service_perfdata_file_template() << "\n";
```
Comment (SylvestreG) [inline src/xpddefault.cc]: leftovers

File: src/xpddefault.cc [hunk 8]
```diff
@@ -18,6 +18,7 @@
 ** <http://www.gnu.org/licenses/>.
 */
 
+#include <iostream>
```
Comment (SylvestreG) [inline src/xpddefault.cc]: leftovers

File: tests/configuration/applier/applier-anomalydetection.cc [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+//  ASSERT_FALSE(serv->get_stalk_on(engine::service::warning));
+//  ASSERT_FALSE(serv->get_stalk_on(engine::service::critical));
+//  ASSERT_FALSE(serv->get_stalk_on(engine::service::unknown));
+//
+//  ASSERT_TRUE(serv->get_notify_on(engine::service::ok));
+//  ASSERT_TRUE(serv->get_notify_on(engine::service::warning));
+//  ASSERT_TRUE(serv->get_notify_on(engine::service::critical));
+//  ASSERT_TRUE(serv->get_notify_on(engine::service::unknown));
+//}
+//
```
Comment (SylvestreG) [inline tests/configuration/applier/applier-anomalydetection.cc]: is this commented code can be uncommented ? if it have no futur use can you remove it ?

File: inc/com/centreon/engine/configuration/applier/scheduler.hh [hunk 10]
```diff
@@ -20,13 +20,17 @@
...
 #define CCE_CONFIGURATION_APPLIER_SCHEDULER_HH
 
+#include <set>
 #include <vector>
 #include "com/centreon/engine/configuration/applier/difference.hh"
+#include "com/centreon/engine/exceptions/error.hh"
 #include "com/centreon/engine/configuration/state.hh"
-#include "com/centreon/engine/namespace.hh"
+//#include "com/centreon/engine/host.hh"
+//#include "com/centreon/engine/service.hh"
```
[the remark this answers is not in the document]
Comment (bouda1) [reply inc/com/centreon/engine/configuration/applier/scheduler.hh]: done

File: src/main.cc (hunk 3 above)
Comment (bouda1) [reply src/main.cc]: Yes. Otherwise you won't have any error log return.

File: tests/configuration/applier/applier-anomalydetection.cc (hunk 9 above)
Comment (bouda1) [reply tests/configuration/applier/applier-anomalydetection.cc]: removed

**Pull request merged**